pkg/ingester/errors.go

```diff
@@ -577,3 +595,38 @@ func handlePushErrorWithHTTPGRPC(err error) error {
 	}
 	return err
 }
+
+func handleReadError(err error) error {
+	var (
+		ingesterErr ingesterError
+		errCode     = codes.Internal
+	)
+	if errors.As(err, &ingesterErr) {
+		switch ingesterErr.errorCause() {
+		case mimirpb.TOO_BUSY:
+			errCode = codes.ResourceExhausted
+		case mimirpb.SERVICE_UNAVAILABLE:
+			errCode = codes.Unavailable
+		}
+	}
+	return newErrorWithStatus(err, errCode)
+}
+
+// handleReadErrorWithHTTPGRPC maps ingesterError objects to an appropriate
+// errorWithStatus, which may contain both HTTP and gRPC error codes.
+// TODO this method is needed only for the backwards compatibility,
+// and should be removed in mimir 2.12.0.
+func handleReadErrorWithHTTPGRPC(err error) error {
+	var (
+		ingesterErr ingesterError
+	)
+	if errors.As(err, &ingesterErr) {
+		switch ingesterErr.errorCause() {
+		case mimirpb.TOO_BUSY:
+			return newErrorWithHTTPStatus(err, http.StatusServiceUnavailable)
+		case mimirpb.SERVICE_UNAVAILABLE:
+			return newErrorWithStatus(err, codes.Unavailable)
+		}
+	}
+	return err
```

Comment:
Is it intentional to not return an `errorWithStatus` when the error cause is unknown? Would it make sense to return `errWithStatus` where the error code is `codes.Internal`?

Reply:
@aknuds1 As it is stated in the function documentation, its purpose is backwads compatibility. My idea was that when `-ingester.return-only-grpc-errors` is false, that the semantics of ingester's read path is the same before and after this PR. That's why in this case we return `err`, to keep the current behavior.

In the case of `handleReadError()` (renamed into `mapReadErrorToErrorWithStatus()`, the error is returned with `codes.Internal` state.

Reply:
OK so old behaviour is to return `err` unadorned, I wasn't sure about this.